PhysLean/Mathematics/VariationalCalculus/IsLocalizedfunctionTransform.lean

```diff
@@ -0,0 +1,317 @@
+/-
+Copyright (c) 2025 Tomas Skrivan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tomas Skrivan, Joseph Tooby-Smith
+-/
+import Mathlib.Analysis.Calculus.BumpFunction.InnerProduct
+import Mathlib.Analysis.Calculus.Gradient.Basic
+import Mathlib.Analysis.InnerProductSpace.Adjoint
+import Mathlib.MeasureTheory.Integral.IntegralEqImproper
+import Mathlib.Analysis.InnerProductSpace.ProdL2
+import Mathlib.Analysis.Calculus.LineDeriv.IntegrationByParts
+
+import PhysLean.Mathematics.VariationalCalculus.Basic
+import PhysLean.Mathematics.InnerProductSpace.Calculus
+import PhysLean.Mathematics.InnerProductSpace.Adjoint
+import PhysLean.Mathematics.Calculus.Divergence
+import PhysLean.Mathematics.Calculus.AdjFDeriv
+import PhysLean.SpaceAndTime.Space.VectorIdentities
+open InnerProductSpace MeasureTheory ContDiff
+
+section
+
+variable
+  {X} [NormedAddCommGroup X]
+  {Y} [NormedAddCommGroup Y]
+  {Z} [NormedAddCommGroup Z]
+  {U}
+  {V}
+  {W}
+
+/-- Function transformation `F` is localizable if the values of the transformed function `F φ` on
+some compact set `K` can depend only on the values of `φ` on some another compact set `L`. -/
+def IsLocalizedFunctionTransform (F : (X → U) → (Y → V)) : Prop :=
+  ∀ (K : Set Y) (_ : IsCompact K), ∃ L : Set X,
+    IsCompact L ∧ ∀ (φ φ' : X → U), (∀ x ∈ L, φ x = φ' x) → ∀ x ∈ K, F φ x = F φ' x
+end
+namespace IsLocalizedFunctionTransform
+
+section
+variable
+  {X} [NormedAddCommGroup X]
+  {Y} [NormedAddCommGroup Y]
+  {Z} [NormedAddCommGroup Z]
+  {U}
+  {V}
+  {W}
+lemma comp {F : (Y → V) → (Z → W)} {G : (X → U) → (Y → V)}
+    (hF : IsLocalizedFunctionTransform F) (hG : IsLocalizedFunctionTransform G) :
+    IsLocalizedFunctionTransform (F ∘ G) := by
+  intro K cK
+  obtain ⟨K', cK', h'⟩ := hF K cK
+  obtain ⟨K'', cK'', h''⟩ := hG K' cK'
+  use K''
+  constructor
+  · exact cK''
+  · intro φ φ' hφ
+    apply h' _ _ (fun _ hx' => h'' _ _ hφ _ hx')
+
+lemma fun_comp {F : (Y → V) → (Z → W)} {G : (X → U) → (Y → V)}
+    (hF : IsLocalizedFunctionTransform F) (hG : IsLocalizedFunctionTransform G) :
+    IsLocalizedFunctionTransform (fun x => F (G x)) := by
+  apply comp hF hG
+
+@[simp]
+lemma id : IsLocalizedFunctionTransform (id : (Y → V) → (Y → V)) := by
+  intro K cK
+  use K
+  constructor
+  · exact cK
+  · intro φ φ' hφ
+    intro x hx
+    simp
+    rw [hφ x hx]
+
```

Comment:
At this point you could formulate 
``` 
lemma fmap (f : Y -> V -> W) (F : (X -> U) -> (Y -> V)) (hF : IsLocalizedFunctionTransform F): IsLocalizedFunctionTransform (fun g y => f y (F g y)) := ...
```
Then proof of `add` would be something like
```
apply fmap (f:= fun _ y => y.1 + y.2) (F:=fun g x => (F g x, G g x))
apply prod hF hG
```

Reply:
But you can leave it as it is